[Workspace] Add documentation for workspace and ACL (#8643)
opensearch-project/documentation-website: _dashboards/workspace/apis.md

---
layout: default
title: Workspaces APIs
parent: Workspace
nav_order: 10
---

> # Workspaces APIs

Reviewer: ```suggestion
# Workspaces APIs
Introduced 2.18
{: .label .label-purple }
```

The Workspaces API provides a set of endpoints for managing workspaces in OpenSearch Dashboards.

## List Workspaces API

You can use the following endpoint to retrieve a list of workspaces.

```json
POST <osd host>:<port>/api/workspaces/_list
```
{% include copy-curl.html %}

The following table lists the available path parameters.

| Parameter | Data type | Required | Description |
| :--- | :--- | :--- | :--- |
| `search` | String | Optional | Query string to filter workspaces using simple query syntax, for example, `simple_query_string`. |
| `searchFields` | Array | Optional | Specifies which fields to perform the search query against. |
| `sortField` | String | Optional | Field name to use for sorting results. |
| `sortOrder` | String | Optional | Specifies ascending or descending sort order. |
| `perPage` | Number | Optional | Number of workspace results per page. |
| `page` | Number | Optional | Page number of results to retrieve. |

> | `permissionModes` | Array | Optional | List of permission moded to filter by. |

Reviewer: ```suggestion
| `permissionModes` | Array | Optional | List of permissions to filter by. |
```

---
layout: default
title: Workspaces APIs
parent: Workspace
nav_order: 10
---

# Workspaces APIs
Introduced 2.18
{: .label .label-purple }

The Workspaces API provides a set of endpoints for managing workspaces in OpenSearch Dashboards.

## List Workspaces API

You can use the following endpoint to retrieve a list of workspaces.

```json
POST <osd host>:<port>/api/workspaces/_list
```
{% include copy-curl.html %}

The following table lists the available path parameters.

| Parameter | Data type | Required | Description |
| :--- | :--- | :--- | :--- |
| `search` | String | Optional | Query string to filter workspaces using simple query syntax, for example, `simple_query_string`. |
| `searchFields` | Array | Optional | Specifies which fields to perform the search query against. |
| `sortField` | String | Optional | Field name to use for sorting results. |
| `sortOrder` | String | Optional | Specifies ascending or descending sort order. |
| `perPage` | Number | Optional | Number of workspace results per page. |

> | `page` | Number | Optional | Page number of results to retrieve. |

Reviewer: Revised to  `The number of pages of results to retrieve`

> | `page` | Number | Optional | Page number of results to retrieve. |

Reviewer: ```suggestion
| `page` | Number | Optional | The number of pages of results to retrieve. |
```

| `attributes` | Object | Required | Defines the workspace attributes. |
| `permissions` | Object | Optional | Specifies the permissions for the workspace. |

#### Example request

```json
PUT api/workspaces/eHVoCJ
{
    "attributes": {
        "name": "test4",
        "description": "test update"
    }
}
```
{% include copy-curl.html %}

The following example response shows a successful API call:

```json
{
    "success": true,
    "result": true
}
```
{% include copy-curl.html %}

## Delete Workspaces API

You can use the following endpoint to delete a workspace:

```json
DELETE <osd host>:<port>/api/workspaces/<id>
```
{% include copy-curl.html %}

The following table lists the available path parameters. All path parameters are required.

| Parameter | Data type | Required | Description |
| :--- | :--- | :--- | :--- |
| `<id>` | String | Required | Identifies the unique workspace to retrieve. |

#### Example request

```json
DELETE api/workspaces/eHVoCJ
```
{% include copy-curl.html %}

The following example response shows a successful API call:

```json
{
    "success": true,
    "result": true
}
```
{% include copy-curl.html %}

> ## Duplicate saved objects Workspaces API

Reviewer: ```suggestion
## Duplicate saved objects Workspaces API
```
```suggestion
## Duplicate Saved Objects Workspaces API
```

The following table lists the attributes of the object in the `objects` parameter.

| Parameter | Data type | Required | Description |
| :--- | :--- | :--- | :--- |
| `type` | String | Required | Defines the saved object classification, such as `index-pattern`, `config`, or `dashboard`. |
| `id` | String | Required | The ID of the saved object. |

#### Example request

```json
POST api/workspaces/_duplicate_saved_objects
{
    "objects": [
        {
            "type": "index-pattern",
            "id": "619cc200-ecd0-11ee-95b1-e7363f9e289d"
        }
    ],
    "targetWorkspace": "9gt4lB"
}
```
{% include copy-curl.html %}

The following example response shows a successful API call:

```json
{
    "successCount": 1,
    "success": true,
    "successResults": [
        {
            "type": "index-pattern",
            "id": "619cc200-ecd0-11ee-95b1-e7363f9e289d",
            "meta": {
                "title": "test*",
                "icon": "indexPatternApp"
            },
            "destinationId": "f4b724fd-9647-4bbf-bf59-610b43a62c75"
        }
    ]
}
```
{% include copy-curl.html %}

> ## Associate saved objects Workspaces API

Reviewer: ```suggestion
## Associate Saved Objects Workspaces API
```

The following table lists the available path parameters.

| Parameter | Data type | Required | Description |
| :--- | :--- | :--- | :--- |
| `workspaceId` | String | Required | Identifies the target workspace for object association. |
| `savedObjects` | Array | Required | Specifies the list of saved objects to be copied. |

The following table lists the attributes of the object in the `objects` parameter.

| Parameter | Data type | Required | Description |
| :--- | :--- | :--- | :--- |
| `type` | String | Required | Defines the saved object classification, such as `index-pattern`, `config`, or `dashboard`. |
| `id` | String | Required | The ID of the saved object. |

#### Example request

```json
POST api/workspaces/_associate
{
    "objects": [
        {
            "type": "index-pattern",
            "id": "619cc200-ecd0-11ee-95b1-e7363f9e289d"
        }
    ],
    "targetWorkspace": "9gt4lB"
}
```
{% include copy-curl.html %}

The following example response shows a successful API call:

```json
{
    "success": true,
    "result": [
        {
            "id": "619cc200-ecd0-11ee-95b1-e7363f9e289d",
        }
    ]
}
```
{% include copy-curl.html %}

> ## Dissociate saved objects Workspaces API

Reviewer: ```suggestion
## Dissociate Saved Objects Workspaces API
```
